File: security/token/php/token_generator.php
<?php

    function generate_token($resource, $generator, $secret_key) {
        /* accepts arguments:
           $resource = a resource on a web server e.g. /path/to/resource?option=33
           $generator = an integer generator value
           $secret_key = the secret key for the HMAC construction.

           Returns the concatenation of the generator and the first 20 characters 
           of the HMAC of $resource using key $secret_key and the SHA1 algorithm.
         */
        $hmac_str = hash_hmac('sha1', $resource, $secret_key);
        return $generator . substr($hmac_str, 0, 20);
    }

    function construct_url_with_token($resource, $generator, $secret) {
        /* construct_url_with token takes arguments:
           $resource = a resource on a web server e.g. /path/to/resource?option=33
           $generator = an integer generator value
           $secret_key = the secret key for the HMAC construction.

           This will append &encoded=token to your resource url
         */
        $token = generate_token($resource, $generator, $secret);
        $new_resource = $resource . "&encoded=" . $token;
        return $new_resource;
    }


    /*
     * This block ensures the code running here only runs 
     * if you execute this file directly using PHP.
     */
    if (!count(debug_backtrace())) {
        /* you can use this snippet in your own code if you
           include 'path/to/token_generator.php';
         */
        $g = 0;
        $key = "aMSyStRoHpHVMflazeGefCofoKoRvDcalsdfkjEuRkbNDhOoekekuFNaMWaYrnprhfm";
        $uri = "/path/to/resource?sessionid=12345&misc=abcde&stime=20081201060100&etime=20201201060100";
        $expected_outcome = "/path/to/resource?sessionid=12345&misc=abcde&stime=20081201060100&etime=20201201060100&encoded=051e88c0261c6a4d83dbc";
        $uri_plus_token = construct_url_with_token($uri, $g, $key);

        /* you don't need this - this is to verify the code produces the 
           expected result */
        print("Result is " . $uri_plus_token . "\n");
        assert($uri_plus_token == $expected_outcome);
    }

?>

File: security/token/python/token_generator.py
"""

This construction allows you to demo code using a preset example
you can use this code directly with:

    from token_generator import sign_url

For general use,
Example URL:
/path/to/protect?stime=20081201060100&etime=20151201060100

Where stime= is the start time, and etime is the end time
(both following yyyymmddHHMMSS format)

Please *note*: for SS authentication, the `stime' and `etime'
arguments in the query string are mandatory as part of the
resource specifier.

Tokens optionally support the use of 'ip' parameter, e.g.:
/path/to/protect?stime=20081201060100&etime=20151201060100&ip=10.0.0.5

"""

import hashlib
import hmac
import argparse

parser = argparse.ArgumentParser(prog="token_generator.py",
    description='''A sample token authentication app for Python''')
parser.add_argument('-u', '--url', nargs=1,
    help='URL to parse (path only)', required=True)
parser.add_argument('-k', '--key', nargs=1,
    help='PSK from the portal', required=True)
params = parser.parse_args()

__all__ = ["generate_token", "sign_url"]


def generate_token(resource, secret_key):
    """ Generate the token based on the resource (URL) and the secret key
    taken from the portal. This is based on the first 20 characters of the
    sha1sum of the secret key and the URL
    """
    hmac_builder = hmac.new(secret_key, resource, hashlib.sha1)
    print hmac_builder.hexdigest()
    return "0%s" % (hmac_builder.hexdigest()[:20])

def sign_url(url, secret):
    """ Return the signed URL with the token added as a new queryString argument
    ("encoded").
    """
    token = generate_token(url, secret)
    return url + "&encoded=" + token

signedUrl = sign_url(params.url[0], params.key[0])
print signedUrl
